File: 1475 final-prices-with-a-special-discount-in-a-shop/solution.cs
public class Solution {
    public int[] FinalPrices(int[] prices) {
        int[] st = new int[prices.Length];
        int top = 0,idx=0;
        int[] ans = new int[prices.Length];
        for(int i=1;i<prices.Length;i++) {
            while(top!=-1 && prices[st[top]] >= prices[i]) {
                ans[st[top]] = prices[st[top--]] - prices[i]; 
            }
            st[++top] = i;
        }
        for(int i=0;i<=top;i++) ans[st[i]] = prices[st[i]];
        return ans;
    }
}

File: 1475 final-prices-with-a-special-discount-in-a-shop/solution.rs
impl Solution {
    pub fn final_prices(prices: Vec<i32>) -> Vec<i32> {
        let n = prices.len();
        let mut ans = vec![0; n];
        let mut stack: Vec<usize> = Vec::new(); 
        for i in 0..n {
            while let Some(&idx) = stack.last() {
                if prices[idx] >= prices[i] {
                    ans[idx] = prices[idx] - prices[i];
                    stack.pop();
                } else {
                    break;
                }
            }
            stack.push(i);
        }
        for idx in stack {
            ans[idx] = prices[idx];
        }

        ans
    }
}

